### ISSUE TITLE
dimension problem

### ISSUE THREAD

**Issue by amy-deng** [yao8839836/text_gcn#10]
Hello,
I read the paper and the code but still have some problems.

The adjacency matrix is present as a graph containing all docs and words. How do you train the weight matrix and apply the weight to your testing set? Since you are supposed to have different dimensions when multiply. Are you using full-batch or stochastic gradient descend? 

Thank you!

**Comment by yao8839836** [comment]
@amy-deng 
Hi, thanks for your question.

Nodes in test set (without labels) are included in training. X has N (N = training docs size + validation docs size + test docs size + vocabulary size) rows and N columns. A^hat also has N rows and N columns. W_0 has N rows and K (the dimension of the first layer embeddings) columns. W_1 has K rows and C (number of classes, also the dimension of the second layer embeddings) columns.

I am using full batch as the original GCN implementation: https://github.com/tkipf/gcn

**Comment by ssdutSqx** [comment]
> After training, the embeddings of test docs nodes are also produced, which can be fed 
> into softmax classifier to predict labels directly.

你好，我是一位图卷积的初学者。有关这里面的矩阵维度我有一个疑问。
原文中给出 Z = softmax(A^hat ReLu(A^hat X W_0) W_1 )
这里在softmax之前的维度应该是 N * C
但是如果要做交叉墒是不是维度应当是 training_size * C （测试集 test_size * C)
希望能解惑 谢谢

**Comment by yao8839836** [comment]
你好，

softmax之前的维度是 N * C，训练时只取N行中的training_size行计算交叉熵，测试时只取N行中的test_size行计算交叉熵。 即只取https://github.com/yao8839836/text_gcn/blob/master/metrics.py中被mask的那些行。

**Comment by ssdutSqx** [comment]
@yao8839836 
你好，不好意思又要打扰了。
Z = softmax(A^hat ReLu(A^hat X W_0) W_1 )做softmax之前是采用mask的方法

假定 A^hat的维度 N * N 其中 N = n_vocab + n_doc，那么在mask的时候n_vocab的部分就完全被去除了。那么之前用PMI计算词语之间共现的部分 n_vocab * n_vocab 有什么用处嘛？

**Comment by yao8839836** [comment]
@ssdutSqx 

你好，n_vocab的部分不参与loss计算，但是为文档之间架起了桥梁，因为GCN本质是邻居节点之间互相传递消息。有label的文档doc 1能将标签信息通过该文档包含的词word 1这座桥传递到其他包含word 1的文档节点doc 2，信息传递路径为doc 1-->word 1-->doc 2; 或者通过与word 1的PMI大于0的其他词节点word 2传递给包含word 2的文档doc 3, 信息传递路径为doc 1--> word 1 --> word 2 --> doc 3。
